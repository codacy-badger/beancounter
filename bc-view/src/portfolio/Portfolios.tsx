import React, { useEffect, useState } from "react";
import { Portfolio } from "../types/beancounter";
import { AxiosError } from "axios";
import logger from "../common/ConfigLogging";
import { axiosBff } from "../common/utils";
import { getBearerToken } from "../keycloak/utils";
import { useKeycloak } from "@react-keycloak/web";
import { Link } from "react-router-dom";

export default function Portfolios(): React.ReactElement {
  const [portfolios, setPortfolios] = useState<Portfolio[]>();
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<AxiosError>();
  const [keycloak] = useKeycloak();
  useEffect(() => {
    const fetchPortfolios = async (config: {
      headers: { Authorization: string };
    }): Promise<void> => {
      setLoading(true);
      logger.debug(">>fetch apiPortfolios");
      await axiosBff()
        .get<Portfolio[]>("/bff/portfolios", config)
        .then(result => {
          logger.debug("<<fetched apiPortfolios");
          setPortfolios(result.data);
        })
        .catch(err => {
          setError(err);
          if (err.response) {
            logger.error("bff error [%s]: [%s]", err.response.status, err.response.data.message);
          }
        });
    };
    fetchPortfolios({
      headers: getBearerToken(keycloak)
    }).finally(() => setLoading(false));
  }, [keycloak]);
  // Render where we are in the initialization process
  if (loading) {
    return <div id="root">Loading...</div>;
  }
  if (error) {
    const { response } = error;
    if (response) {
      const { data: errData, status } = response;
      if (status === 401) {
        return <div>Please log in</div>;
      }
      logger.error("Error: %s", errData.message);
      return <div>{errData.message}</div>;
    }
    return <div>Unknown error</div>;
  }
  if (portfolios) {
    return (
      <div className="page-box">
        <table className={"table is-striped is-hoverable"}>
          <tbody>
            {portfolios.map(portfolio => (
              <tr key={portfolio.id}>
                <td align={"left"}>
                  <Link to={`/holdings/${portfolio.code}`}>{portfolio.code}</Link>
                </td>
                <td align={"left"}>{portfolio.name}</td>
                <td>
                  {portfolio.currency.symbol}
                  {portfolio.currency.code}
                </td>
                <td>
                  {portfolio.base.symbol}
                  {portfolio.base.code}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
  return <div id="root">You have no portfolios - create one?</div>;
}
