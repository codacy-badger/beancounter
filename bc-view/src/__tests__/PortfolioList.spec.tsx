import React from "react";
import { cleanup, render, waitForElement } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import nock from "nock";
import Portfolios from "../portfolio/Portfolios";
import { MemoryRouter } from "react-router";

afterEach(cleanup);

const bff = "http://localhost";
nock(bff, {
  reqheaders: {
    authorization: "Bearer undefined"
  }
})
  .get("/bff/portfolios")
  .replyWithFile(200, __dirname + "/contracts/portfolios.json", {
    "Access-Control-Allow-Origin": "*",
    "Content-type": "application/json"
  });

describe("<Portfolios />", () => {
  it("should match snapshot", async () => {
    const { getByText, container } = render(
      <MemoryRouter initialEntries={["/"]} keyLength={0}>
        <Portfolios />
      </MemoryRouter>
    );
    await waitForElement(() => getByText("TEST"));
    expect(nock.isDone());
    expect(container).toMatchSnapshot();
  });
});
