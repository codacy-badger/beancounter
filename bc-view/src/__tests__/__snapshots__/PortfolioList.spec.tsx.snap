// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portfolios /> should match snapshot 1`] = `
<div>
  <div>
    <section
      class="page-box hero is-primary is-fullheight has-background-light"
    >
      <div
        class="hero-body"
      >
        <div
          class="container"
        >
          <div
            class="columns is-centered"
          >
            <table
              class="table is-striped is-hoverable"
            >
              <thead>
                <tr>
                  <th>
                    Code
                  </th>
                  <th>
                    Name
                  </th>
                  <th>
                    Report
                  </th>
                  <th>
                    Base
                  </th>
                  <th>
                    Actions
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    <a
                      href="/holdings/TEST"
                    >
                      TEST
                    </a>
                  </td>
                  <td>
                    Test Portfolio
                  </td>
                  <td>
                    â‚¬
                    EUR
                  </td>
                  <td>
                    $
                    USD
                  </td>
                  <td>
                    <a
                      class="fa fa-edit"
                      href="/portfolio/Kdf3jfUfSYqXV_S4Myc2wA"
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
